rfd/001-tcp-load-balancer-design-doc.md

```diff
@@ -0,0 +1,119 @@
+
+```
+clientID := uuid.NewV5(loadBalancerID, client.TLSCertificate().String())
+```
+
+Client connection counts can be stored in a `sync.Map` structure for quick modification, and safety across concurrent goroutines. Regarding this map choice, the [`sync` package documentation](https://pkg.go.dev/sync#Map) states the following: 
+> "The Map type is optimized for two common use cases: (1) when the entry for a given key is only ever written once but read many times, as in caches that only grow, or (2) when multiple goroutines read, write, and overwrite entries for disjoint sets of keys. In these two cases, use of a Map may significantly reduce lock contention compared to a Go map paired with a separate Mutex or RWMutex."
+
+Our use-case follows the #2 reason for leveraging `sync.Map` over a standard map with a Mutex; our disjoint sets of keys will be the unique client ids.
+
+
+### Upstream Authorization Management
+
+The load balancer will employ a simple authorization scheme which will deny access to specific upstreams for certain clients. This scheme will be statically defined in code. Once the load balancer has determined that the client is authenticated and within client connection limit (and prior to selecting the least connections host) it will filter out unauthorized hosts for that client. To simply demonstrate this capability, we store a map of `clientID`s (based on client certificate) to disallowed `hostID`s; because host configuration is outside the scope of this project, the order in which the host was registered with the load balancer will be used as the `hostID`.
+
+### Request Forwarder
+
+Incoming packets will be forwarded after checking 3 criterion:
+* Client must be authenticated.
+* Client must be within connection limit.
+* Client must be authorized to access at least one healthy host.
+
+Once these requirements are satisfied, the request will be forwarded to an authorized host with the least connections.
+
+The least connection tracking will prioritize simplicity over performance for this project. Each new connection that is made will increment a host's connection count, and each completed connection will decrement the host's connection count. Each incoming request will iterate through all authorized hosts to find the one with the least active connections. A more performant solution could later be devised which may not have to iterate over every authorized host for every request (likely leveraging an async queue). One additional shortcoming of this simple approach is that a large batch of requests which arrive at exact same time could be routed to the exact same host, but this initial project is not planning to handle that edge case.
```

Comment:
> One additional shortcoming of this simple approach is that a large batch of requests which arrive at exact same time could be routed to the exact same host

Do you see any way to avoid this ?

Reply:
Yes, the host selection process could leverage a mutex to lock access to new connections until the connection count has been increased. This would ensure that connection counts are accurate for each new connection. 

### Locking Example (pseudo code)
```
mutex.Lock()
host := lb.LeastConnections()
host.Connect()
host.IncrementConnectionCount()
mutex.Unlock()
```
The drawback would be an increase to latency while selecting hosts.

Reply:
Another option, without locking connections, could work in the case of equal connections across all hosts (probably only useful when traffic patterns involve large simultaneous spikes after sitting at 0 connections), would involve randomly selecting the starting host (rather than starting with the first host) prior to the selection loop. That would ensure a random distribution of simultaneous load, since each incoming connection starts a separate goroutine, each routine would pick a random number, and no hosts would have less connections than the starting host.

### Random Starting Host Example (only useful when connection are equal)
```
	host := l.hosts[rand.Intn(len(hosts))] // changed from 'host := l.hosts[0]`

	for i, h := range l.hosts {
		if h.ConnectionCount() < host.ConnectionCount() {
			host = h
		}
	}

	return host
```
The drawback to this approach is that it's useless if existing connection counts are unequal.

Reply:
Thanks for the explanation and attaching examples: 

Regarding "Locking Example" 
```
mutex.Lock()
host := lb.LeastConnections()
host.Connect()
host.IncrementConnectionCount()
mutex.Unlock()
```

Does the host.Connect() call needs to be done under a lock ?

Reply:
No, it does not need to be done under a lock, and having it outside the lock will significantly reduce latency. 

The trade-off of excluding the connection from the lock is that we won't know whether or not the connection succeeded until after we have incremented the connection count; if we're momentarily off by 1 on an unhealthy host's connection count, it should have no material impact on LB functionality. Seems like a good tradeoff to make.

Should I update the design doc so that the implementation plan includes connection count locking?